### PR TITLE
Improve consistency in the C++ interface

### PR DESCRIPTION
**Pull request by cubicYYY**
These changes aim to improve the consistency of the C++ API between the 32-bit version and the 64-bit version.

Related issues: #629 

Changes:

- Rename `RoaringSetBitForwardIterator` to `RoaringSetBitBiDirectionalIterator` (the old one is now deprecated).
- Add methods: `Roaring::clear` `Roaring::isFull` `Roaring::containsRangeClosed`.
- Revise `move_equalorlarger` methods in iterators.
- Add the `api::roaring_bitmap_flip_inplace_closed` method.
- Add the `api::roaring_bitmap_contains_range_closed` method.
- Add the `api::roaring_bitmap_flip_closed` method.
- Add the `api::roaring_bitmap_range_cardinality_closed` method.
- Other minor changes.

If merged, only the following methods remain that are only implemented in the 32-bit version(in class `Roaring`) but not in the 64-bit version(in class `Roaring64Map`):

- [ ]  `containsRange`
- [ ]  `rangeUint64Array`
- [ ]  `and_cardinality`
- [ ]  `or_cardinality`
- [ ]  `andnot_cardinality`
- [ ]  `xor_cardinality`
- [ ]  `jaccard_index`
- [ ]  `rank_many`

Also, the methods mentioned in #549 need to be further implemented.

### CODE REVIEW

File: cpp/roaring64map.hh [hunk 1]
```diff
@@ -844,6 +840,8 @@ class Roaring64Map {
             });
     }
 
+    // TODO: implement `rangeUint64Array`
```
Comment (lemire) [inline cpp/roaring64map.hh:843]:
```suggestion
```

Same comment here, let us not add too many todos in the code when we have the option of creating GH issues.

File: cpp/roaring64map.hh [hunk 2]
```diff
@@ -1722,8 +1720,8 @@ class Roaring64MapSetBitBiDirectionalIterator {
     typedef int64_t difference_type;
     typedef Roaring64MapSetBitBiDirectionalIterator type_of_iterator;
 
-    Roaring64MapSetBitBiDirectionalIterator(const Roaring64Map &parent,
-                                            bool exhausted = false)
+    explicit Roaring64MapSetBitBiDirectionalIterator(const Roaring64Map &parent,
```
Comment (lemire) [inline cpp/roaring64map.hh:1723]: I don't mind the explicit but it is potentially a breaking change. Can you justify the explicit?

We would rather not break people's code.

File: src/roaring.c [hunk 3]
```diff
@@ -2005,20 +2006,26 @@ static void inplace_fully_flip_container(roaring_array_t *x1_arr, uint16_t hb) {
 roaring_bitmap_t *roaring_bitmap_flip(const roaring_bitmap_t *x1,
                                       uint64_t range_start,
                                       uint64_t range_end) {
-    if (range_start >= range_end) {
+    if (range_start >= range_end || range_start > (uint64_t)UINT_MAX + 1)
```
Comment (lemire) [inline src/roaring.c:2009]:
```suggestion
    if (range_start >= range_end || range_start > (uint64_t)UINT32_MAX + 1)
```

In theory, UINT_MAX could be UINT64_MAX (I don't think this actually happens, but it might).

File: src/roaring.c [hunk 4]
```diff
@@ -2059,17 +2066,23 @@ roaring_bitmap_t *roaring_bitmap_flip(const roaring_bitmap_t *x1,
 
 void roaring_bitmap_flip_inplace(roaring_bitmap_t *x1, uint64_t range_start,
                                  uint64_t range_end) {
-    if (range_start >= range_end) {
+    if (range_start >= range_end || range_start > (uint64_t)UINT_MAX + 1)
```
Comment (lemire) [inline src/roaring.c:2069]:
```suggestion
    if (range_start >= range_end || range_start > (uint64_t)UINT32_MAX + 1)
```

File: src/roaring.c [hunk 5]
```diff
@@ -2827,15 +2840,25 @@ bool roaring_bitmap_contains(const roaring_bitmap_t *r, uint32_t val) {
  */
 bool roaring_bitmap_contains_range(const roaring_bitmap_t *r,
                                    uint64_t range_start, uint64_t range_end) {
-    if (range_end >= UINT64_C(0x100000000)) {
-        range_end = UINT64_C(0x100000000);
-    }
-    if (range_start >= range_end)
+    if (range_start >= range_end || range_start > (uint64_t)UINT_MAX + 1)
```
Comment (lemire) [inline src/roaring.c:2843]:
```suggestion
    if (range_start >= range_end || range_start > (uint64_t)UINT32_MAX + 1)
```

**Comment by lemire** [comment]
Looks good to me. I invited others to comment.

Let us keep the breaking changes to a minimum. Deprecation is fine.

File: cpp/roaring64map.hh (hunk 2 above)
Comment (cubicYYY) [reply cpp/roaring64map.hh:1723]: Sure. This is added since the corresponding constructor in `Roaring` comes with an _explicit_, hence an _explicit_ here could give us more symmetry.
On the other hand, you are right, backward compatibility matters.
So I'm now neutral about this change and waiting for more points of view.

File: cpp/roaring64map.hh [hunk 6]
```diff
@@ -493,6 +493,8 @@ class Roaring64Map {
         return iter->second.contains(lowBytes(x));
     }
 
+    // TODO: implement `containsRange`
```
[the remark this answers is not in the document]
Comment (cubicYYY) [reply cpp/roaring64map.hh:496]: Sure. I would like to implement these along with a `...Closed` version in this PR.

[the remark this answers is not in the document]
Comment (cubicYYY) [reply cpp/roaring64map.hh:496]: My mind changed. I found that it should be handled in another issue. They require new tests. Let's focus on consistency instead of functionalities in this PR.

[the remark this answers is not in the document]
Comment (lemire) [reply cpp/roaring64map.hh:496]: It is probably better to have more PRs. Small PRs are easier to assess.

File: cpp/roaring.hh [hunk 7]
```diff
@@ -39,7 +40,10 @@ A C++ header for Roaring Bitmaps.
 
 namespace roaring {
 
-class RoaringSetBitForwardIterator;
+class RoaringSetBitBiDirectionalIterator;
+
+/** DEPRECATED, use `RoaringSetBitBiDirectionalIterator`. */
+using RoaringSetBitForwardIterator = RoaringSetBitBiDirectionalIterator;
```
Comment (mapleFU) [inline cpp/roaring.hh:46]: Can we use `CROARING_DEPRECATED`?

File: cpp/roaring.hh [hunk 8]
```diff
@@ -91,6 +95,16 @@ class Roaring {
         addMany(l.size(), l.begin());
     }
 
+    /**
+     * Construct a roaring object by taking control of a malloc()'d C struct.
```
Comment (mapleFU) [inline cpp/roaring.hh:99]: This is just a position moving?

Comment (cubicYYY) [reply cpp/roaring.hh:99]: Yes, to keep the same order as the 32-bit version.

File: cpp/roaring.hh (hunk 7 above)
Comment (cubicYYY) [reply cpp/roaring.hh:46]: I'd like to, but:
1. currently, the `Roaring64MapSetBitForwardIterator` is not marked as deprecated.
2. `CROARING_DEPRECATED` expands to `__attribute__((deprecated))` for GNUC and Clang, so `using xxx CROARING_DEPRECATED = xxx;` works. However, for MSVC, it expands to `__declspec(deprecated)`, which won't work in this format.

File: cpp/roaring64map.hh (hunk 2 above)
Comment (lemire) [reply cpp/roaring64map.hh:1721]: @cubicYYY We can't assume that CRoaring is only used by projects on GitHub. So if we check and find something then we know that it is in use, but if we don't find anything, we don't know whether it is in use.

Comment (cubicYYY) [reply cpp/roaring64map.hh:1721]: @lemire That makes sense. If backward compatibility is a must, what about adding a brief comment instead of actually making it `explicit`?

Comment (lemire) [reply cpp/roaring64map.hh:1721]: Honestly, I recommend just dropping the issue and removing the explicit. The motivation seems weak:

> This is added since the corresponding constructor in Roaring comes with an explicit, hence an explicit here could give us more symmetry.

Can we come up with one example where this change helps anyone?

If a change is not helpful, and if it is purely for aesthetic, then I recommend dropping it. Even if there is a very tiny chance that it breaks someone's code, it is worse than an aesthetic issue nobody is ever going to notice in the real world.

In other words, it seems to me that this is not worth changing.

Comment (cubicYYY) [reply cpp/roaring64map.hh:1721]: Roger!

File: benchmarks/containsmulti_benchmark.c [hunk 9]
```diff
@@ -19,7 +19,7 @@ void contains_multi_via_contains(roaring_bitmap_t* bm, const uint32_t* values,
 
 void contains_multi_bulk(roaring_bitmap_t* bm, const uint32_t* values,
                          bool* results, const size_t count) {
-    roaring_bulk_context_t context = {0, 0};
+    roaring_bulk_context_t context = {0, 0, 0, 0};
```
Comment (SLieve) [inline benchmarks/containsmulti_benchmark.c:22]: You can use `CROARING_ZERO_INITIALIZER` from `portabilty.h`

File: cpp/roaring64map.hh (hunk 6 above)
[the remark this answers is not in the document]
Comment (SLieve) [reply cpp/roaring64map.hh:496]: This should be removed then.

File: cpp/roaring64map.hh [hunk 10]
```diff
@@ -1712,6 +1708,9 @@ class Roaring64Map {
 
 /**
  * Used to go through the set bits. Not optimally fast, but convenient.
+ *
+ * strongly recommend not to implicitly construct this iterator:
+ * implicit construction may be prohibited in the future.
```
Comment (SLieve) [inline cpp/roaring64map.hh:1713]: I think it's possible that this will happen but not very likely, also the harm of implicit construction seems relatively small in the case of the iterator.

File: src/roaring.c [hunk 11]
```diff
@@ -2005,20 +2013,26 @@ static void inplace_fully_flip_container(roaring_array_t *x1_arr, uint16_t hb) {
 roaring_bitmap_t *roaring_bitmap_flip(const roaring_bitmap_t *x1,
                                       uint64_t range_start,
                                       uint64_t range_end) {
-    if (range_start >= range_end) {
+    if (range_start >= range_end || range_start > (uint64_t)UINT32_MAX + 1)
+        return roaring_bitmap_copy(x1);
```
Comment (SLieve) [inline src/roaring.c:2017]: Nit/style preference: most of the codebase uses parentheses even for single-statement if expressions. Same elsewhere in this PR.

**Review by SLieve — APPROVED** [review]
Looks good, some small nits.

File: src/roaring.c (hunk 11 above)
Comment (lemire) [reply src/roaring.c:2017]: @SLieve Maybe you mean curly brackets? As in `if(condition) { dothis }`  as opposed to if(condition)  dothis `.

Comment (SLieve) [reply src/roaring.c:2017]: Err yes, my mistake!

**Comment by lemire** [comment]
Running tests with the latest changes.
